File: protocol_codec.ts
/**
 * Auto-Announce UDP Protocol v1 codec (TypeScript).
 *
 * Dependency-free, intended for browser/Node/firmware tooling.
 */

export const MAGIC = 0xaa55
export const VERSION = 1
export const HEADER_SIZE = 22

export const MSG_AUDIO = 0x01
export const MSG_TELEMETRY = 0x02
export const MSG_CONTROL = 0x03

export const FLAG_START = 0x01
export const FLAG_END = 0x02
export const FLAG_RESYNC = 0x04

export const STATUS_STREAM_PRESENT = 0x01
export const STATUS_TIMEOUT_MUTE = 0x02
export const STATUS_CLIPPING = 0x04
export const STATUS_CALIBRATION = 0x08
export const STATUS_LOW_BATTERY = 0x10
export const STATUS_LOW_TEMP = 0x20

export const ENC_PCM16_LE = 0x01

export interface PacketHeader {
  magic: number
  version: number
  msgType: number
  zoneId: number
  streamId: number
  seq: number
  timestamp: number
  flags: number
  crc16: number
  payloadLen: number
}

export interface AudioPayload {
  encoding: number
  channels: number
  sampleRateHz: number
  frameSamples: number
  audio: Uint8Array
}

export interface TelemetryPayload {
  status: number
  rssi: number
  ambientRms: number
  bufferMs: number
  jitterMs: number
  packetLossPpm: number
  lastSeq: number
  streamId: number
}

export interface ControlPayload {
  op: number
  zoneId: number
  argU32: number
}

export type PacketPayload = AudioPayload | TelemetryPayload | ControlPayload | Uint8Array

export interface ParsedPacket {
  header: PacketHeader
  payload: PacketPayload
}

export function crc16Ccitt(data: Uint8Array, poly = 0x1021, init = 0xffff): number {
  let crc = init
  for (const byte of data) {
    crc ^= byte << 8
    for (let bit = 0; bit < 8; bit++) {
      if (crc & 0x8000) {
        crc = ((crc << 1) ^ poly) & 0xffff
      } else {
        crc = (crc << 1) & 0xffff
      }
    }
  }
  return crc
}

function writeHeader(header: Omit<PacketHeader, "payloadLen" | "crc16" | "magic" | "version"> & {
  payloadLen: number
  crc16?: number
}, buf: Uint8Array, includeCrc = false) {
  const dv = new DataView(buf.buffer, buf.byteOffset, buf.byteLength)
  let offset = 0
  dv.setUint16(offset, MAGIC, true); offset += 2
  dv.setUint8(offset, VERSION); offset += 1
  dv.setUint8(offset, header.msgType); offset += 1
  dv.setUint16(offset, header.zoneId & 0xffff, true); offset += 2
  dv.setUint16(offset, header.streamId & 0xffff, true); offset += 2
  dv.setUint32(offset, header.seq >>> 0, true); offset += 4
  dv.setUint32(offset, header.timestamp >>> 0, true); offset += 4
  dv.setUint8(offset, header.flags & 0xff); offset += 1
  dv.setUint8(offset, 0); offset += 1
  dv.setUint16(offset, includeCrc ? (header.crc16 ?? 0) : 0, true); offset += 2
  dv.setUint16(offset, header.payloadLen & 0xffff, true)
}

export function packAudioPacket(
  zoneId: number,
  streamId: number,
  seq: number,
  timestamp: number,
  payload: AudioPayload,
  flags = 0,
  includeCrc = false
): Uint8Array {
  const payloadBytes = new Uint8Array(8 + payload.audio.length)
  const dv = new DataView(payloadBytes.buffer)
  let o = 0
  dv.setUint8(o, payload.encoding); o += 1
  dv.setUint8(o, payload.channels); o += 1
  dv.setUint16(o, payload.sampleRateHz, true); o += 2
  dv.setUint16(o, payload.frameSamples, true); o += 2
  dv.setUint16(o, payload.audio.length, true); o += 2
  payloadBytes.set(payload.audio, o)

  const packet = new Uint8Array(HEADER_SIZE + payloadBytes.length)
  const crc = includeCrc ? crc16Ccitt(payloadBytes) : 0
  writeHeader({ msgType: MSG_AUDIO, zoneId, streamId, seq, timestamp, flags, payloadLen: payloadBytes.length, crc16: crc }, packet, includeCrc)
  packet.set(payloadBytes, HEADER_SIZE)
  return packet
}

export function packTelemetryPacket(
  zoneId: number,
  streamId: number,
  seq: number,
  timestamp: number,
  payload: TelemetryPayload,
  flags = 0,
  includeCrc = false
): Uint8Array {
  const payloadBytes = new Uint8Array(18)
  const dv = new DataView(payloadBytes.buffer)
  let o = 0
  dv.setUint8(o, payload.status); o += 1
  dv.setUint8(o, payload.rssi); o += 1
  dv.setUint16(o, payload.ambientRms, true); o += 2
  dv.setUint16(o, payload.bufferMs, true); o += 2
  dv.setUint16(o, payload.jitterMs, true); o += 2
  dv.setUint16(o, payload.packetLossPpm, true); o += 2
  dv.setUint32(o, payload.lastSeq >>> 0, true); o += 4
  dv.setUint16(o, payload.streamId, true); o += 2
  dv.setUint16(o, 0, true)

  const packet = new Uint8Array(HEADER_SIZE + payloadBytes.length)
  const crc = includeCrc ? crc16Ccitt(payloadBytes) : 0
  writeHeader(
    { msgType: MSG_TELEMETRY, zoneId, streamId, seq, timestamp, flags, payloadLen: payloadBytes.length, crc16: crc },
    packet,
    includeCrc
  )
  packet.set(payloadBytes, HEADER_SIZE)
  return packet
}

export function packControlPacket(
  zoneId: number,
  streamId: number,
  seq: number,
  timestamp: number,
  payload: ControlPayload,
  flags = 0,
  includeCrc = false
): Uint8Array {
  const payloadBytes = new Uint8Array(8)
  const dv = new DataView(payloadBytes.buffer)
  let o = 0
  dv.setUint8(o, payload.op); o += 1
  dv.setUint16(o, payload.zoneId, true); o += 2
  dv.setUint8(o, 0); o += 1
  dv.setUint32(o, payload.argU32 >>> 0, true)

  const packet = new Uint8Array(HEADER_SIZE + payloadBytes.length)
  const crc = includeCrc ? crc16Ccitt(payloadBytes) : 0
  writeHeader({ msgType: MSG_CONTROL, zoneId, streamId, seq, timestamp, flags, payloadLen: payloadBytes.length, crc16: crc }, packet, includeCrc)
  packet.set(payloadBytes, HEADER_SIZE)
  return packet
}

export function parsePacket(raw: Uint8Array): ParsedPacket {
  if (raw.byteLength < HEADER_SIZE) {
    throw new Error("packet too short for header")
  }

  const dv = new DataView(raw.buffer, raw.byteOffset, raw.byteLength)
  let offset = 0
  const magic = dv.getUint16(offset, true); offset += 2
  if (magic !== MAGIC) {
    throw new Error(`invalid magic: 0x${magic.toString(16)}`)
  }

  const version = dv.getUint8(offset); offset += 1
  if (version !== VERSION) {
    throw new Error(`unexpected version: ${version}`)
  }

  const msgType = dv.getUint8(offset); offset += 1
  const zoneId = dv.getUint16(offset, true); offset += 2
  const streamId = dv.getUint16(offset, true); offset += 2
  const seq = dv.getUint32(offset, true); offset += 4
  const timestamp = dv.getUint32(offset, true); offset += 4
  const flags = dv.getUint8(offset); offset += 1
  offset += 1 // reserved
  const crc16 = dv.getUint16(offset, true); offset += 2
  const payloadLen = dv.getUint16(offset, true); offset += 2

  if (HEADER_SIZE + payloadLen !== raw.byteLength) {
    throw new Error("payload_len mismatch with datagram length")
  }

  const payload = new Uint8Array(raw.buffer, raw.byteOffset + HEADER_SIZE, payloadLen)
  if (crc16 !== 0 && crc16Ccitt(payload) !== crc16) {
    throw new Error("crc16 mismatch")
  }

  const header: PacketHeader = {
    magic,
    version,
    msgType,
    zoneId,
    streamId,
    seq,
    timestamp,
    flags,
    crc16,
    payloadLen
  }

  if (msgType === MSG_AUDIO) {
    if (payloadLen < 8) throw new Error("audio payload too short")
    const d = new DataView(payload.buffer, payload.byteOffset, payload.byteLength)
    let po = 0
    const encoding = d.getUint8(po); po += 1
    const channels = d.getUint8(po); po += 1
    const sampleRateHz = d.getUint16(po, true); po += 2
    const frameSamples = d.getUint16(po, true); po += 2
    const audioBytes = d.getUint16(po, true); po += 2
    if (payload.byteLength !== 8 + audioBytes) throw new Error("audio payload length mismatch")
    const audio = payload.slice(8)
    return {
      header,
      payload: {
        encoding,
        channels,
        sampleRateHz,
        frameSamples,
        audio
      }
    }
  }

  if (msgType === MSG_TELEMETRY) {
    if (payloadLen < 18) throw new Error("telemetry payload too short")
    const d = new DataView(payload.buffer, payload.byteOffset, payload.byteLength)
    let po = 0
    const status = d.getUint8(po); po += 1
    const rssi = d.getUint8(po); po += 1
    const ambientRms = d.getUint16(po, true); po += 2
    const bufferMs = d.getUint16(po, true); po += 2
    const jitterMs = d.getUint16(po, true); po += 2
    const packetLossPpm = d.getUint16(po, true); po += 2
    const lastSeq = d.getUint32(po, true); po += 4
    const seenStreamId = d.getUint16(po, true); po += 2
    return {
      header,
      payload: {
        status,
        rssi,
        ambientRms,
        bufferMs,
        jitterMs,
        packetLossPpm,
        lastSeq,
        streamId: seenStreamId
      }
    }
  }

  if (msgType === MSG_CONTROL) {
    if (payloadLen < 8) throw new Error("control payload too short")
    const d = new DataView(payload.buffer, payload.byteOffset, payload.byteLength)
    const op = d.getUint8(0)
    const payloadZoneId = d.getUint16(1, true)
    const argU32 = d.getUint32(4, true)
    return {
      header,
      payload: {
        op,
        zoneId: payloadZoneId,
        argU32
      }
    }
  }

  return { header, payload }
}

export function demo(): void {
  const audio = {
    encoding: ENC_PCM16_LE,
    channels: 1,
    sampleRateHz: 48_000,
    frameSamples: 480,
    audio: new Uint8Array(960)
  }
  const packet = packAudioPacket(2, 7, 1, 0, audio, FLAG_START, true)
  const decoded = parsePacket(packet)
  const p = decoded.payload as AudioPayload
  console.log(`Demo bytes=${packet.byteLength}`)
  console.log(`Zone=${decoded.header.zoneId} Seq=${decoded.header.seq} type=${decoded.header.msgType}`)
  console.log(`Samples=${p.frameSamples} audio_bytes=${p.audio.length}`)
}

File: protocol_codec.py
#!/usr/bin/env python3
"""
Auto-Announce UDP Protocol v1 codec.

No external dependencies. Intended as a validation stub for app/firmware work.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from typing import Dict, Any


MAGIC = 0xAA55
VERSION = 1

MSG_AUDIO = 0x01
MSG_TELEMETRY = 0x02
MSG_CONTROL = 0x03

FLAG_START = 0x01
FLAG_END = 0x02
FLAG_RESYNC = 0x04

STATUS_STREAM_PRESENT = 0x01
STATUS_TIMEOUT_MUTE = 0x02
STATUS_CLIPPING = 0x04
STATUS_CALIBRATION = 0x08
STATUS_LOW_BATTERY = 0x10
STATUS_LOW_TEMP = 0x20

ENC_PCM16_LE = 0x01


def crc16_ccitt(data: bytes, poly: int = 0x1021, init: int = 0xFFFF) -> int:
    crc = init
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            crc = ((crc << 1) ^ poly) & 0xFFFF if (crc & 0x8000) else (crc << 1) & 0xFFFF
    return crc & 0xFFFF


@dataclass
class Packet:
    msg_type: int
    zone_id: int
    stream_id: int
    seq: int
    timestamp: int
    flags: int = 0
    payload: bytes = b""

    def pack(self, version: int = VERSION, include_crc: bool = False) -> bytes:
        payload_len = len(self.payload)
        crc = crc16_ccitt(self.payload) if include_crc else 0
        hdr = struct.pack(
            "<HBBHHIIBBHH",
            MAGIC,
            version,
            self.msg_type,
            self.zone_id & 0xFFFF,
            self.stream_id & 0xFFFF,
            self.seq & 0xFFFFFFFF,
            self.timestamp & 0xFFFFFFFF,
            self.flags & 0xFF,
            0,
            crc & 0xFFFF,
            payload_len & 0xFFFF,
        )
        return hdr + self.payload


@dataclass
class AudioPayload:
    encoding: int
    channels: int
    sample_rate_hz: int
    frame_samples: int
    audio: bytes

    def pack(self) -> bytes:
        return struct.pack("<BBHHH", self.encoding, self.channels, self.sample_rate_hz, self.frame_samples, len(self.audio)) + self.audio

    @classmethod
    def parse(cls, data: bytes) -> "AudioPayload":
        if len(data) < 8:
            raise ValueError("audio payload too short")
        encoding, channels, sample_rate_hz, frame_samples, audio_bytes = struct.unpack("<BBHHH", data[:8])
        if len(data) != 8 + audio_bytes:
            raise ValueError("audio payload length mismatch")
        audio = data[8:]
        return cls(encoding, channels, sample_rate_hz, frame_samples, audio)


@dataclass
class TelemetryDebug:
    flags: int = 0
    receiver_pps_x10: int = 0
    total_packets: int = 0
    lost_packets: int = 0
    late_packets: int = 0
    playback_buffer_frames: int = 0
    playback_underflows: int = 0
    playback_drops: int = 0
    playback_broken_pipes: int = 0
    playback_exit_code: int = -1
    parse_errors: int = 0

    def pack(self) -> bytes:
        return struct.pack(
            "<HBBHIIIHIIHhH",
            0x4441,
            1,
            self.flags & 0xFF,
            self.receiver_pps_x10 & 0xFFFF,
            self.total_packets & 0xFFFFFFFF,
            self.lost_packets & 0xFFFFFFFF,
            self.late_packets & 0xFFFFFFFF,
            self.playback_buffer_frames & 0xFFFF,
            self.playback_underflows & 0xFFFFFFFF,
            self.playback_drops & 0xFFFFFFFF,
            self.playback_broken_pipes & 0xFFFF,
            max(-32768, min(32767, self.playback_exit_code)),
            self.parse_errors & 0xFFFF,
        )

    @classmethod
    def parse(cls, data: bytes) -> "TelemetryDebug | None":
        if len(data) < 34:
            return None
        magic, version = struct.unpack("<HB", data[:3])
        if magic != 0x4441 or version != 1:
            return None
        (
            _magic,
            _version,
            flags,
            receiver_pps_x10,
            total_packets,
            lost_packets,
            late_packets,
            playback_buffer_frames,
            playback_underflows,
            playback_drops,
            playback_broken_pipes,
            playback_exit_code,
            parse_errors,
        ) = struct.unpack("<HBBHIIIHIIHhH", data[:34])
        return cls(
            flags,
            receiver_pps_x10,
            total_packets,
            lost_packets,
            late_packets,
            playback_buffer_frames,
            playback_underflows,
            playback_drops,
            playback_broken_pipes,
            playback_exit_code,
            parse_errors,
        )


@dataclass
class TelemetryPayload:
    status: int
    rssi: int
    ambient_rms: int
    buffer_ms: int
    jitter_ms: int
    packet_loss_ppm: int
    last_seq: int
    stream_id: int
    debug: TelemetryDebug | None = None

    def pack(self) -> bytes:
        base = struct.pack(
            "<BBHHHHIHH",
            self.status & 0xFF,
            self.rssi & 0xFF,
            self.ambient_rms & 0xFFFF,
            self.buffer_ms & 0xFFFF,
            self.jitter_ms & 0xFFFF,
            self.packet_loss_ppm & 0xFFFF,
            self.last_seq & 0xFFFFFFFF,
            self.stream_id & 0xFFFF,
            0,
        )
        return base + (self.debug.pack() if self.debug else b"")

    @classmethod
    def parse(cls, data: bytes) -> "TelemetryPayload":
        if len(data) < 18:
            raise ValueError("telemetry payload too short")
        status, rssi, ambient_rms, buffer_ms, jitter_ms, packet_loss_ppm, last_seq, stream_id, _reserved = struct.unpack("<BBHHHHIHH", data[:18])
        debug = TelemetryDebug.parse(data[18:]) if len(data) >= 52 else None
        return cls(status, rssi, ambient_rms, buffer_ms, jitter_ms, packet_loss_ppm, last_seq, stream_id, debug)


@dataclass
class ControlPayload:
    op: int
    zone_id: int
    arg_u32: int

    def pack(self) -> bytes:
        return struct.pack("<BHBI", self.op, self.zone_id & 0xFFFF, 0, self.arg_u32 & 0xFFFFFFFF)

    @classmethod
    def parse(cls, data: bytes) -> "ControlPayload":
        if len(data) < 8:
            raise ValueError("control payload too short")
        op = data[0]
        zone_id = struct.unpack("<H", data[1:3])[0]
        arg_u32 = struct.unpack("<I", data[4:8])[0]
        return cls(op, zone_id, arg_u32)


def parse_packet(raw: bytes) -> Dict[str, Any]:
    if len(raw) < 22:
        raise ValueError("packet too short for header")

    magic, version, msg_type, zone_id, stream_id, seq, timestamp, flags, _reserved, crc16, payload_len = struct.unpack(
        "<HBBHHIIBBHH",
        raw[:22]
    )
    if magic != MAGIC:
        raise ValueError(f"invalid magic: 0x{magic:04X}")
    if payload_len + 22 != len(raw):
        raise ValueError("payload_len mismatch with datagram length")

    payload = raw[22:]
    if crc16 and crc16_ccitt(payload) != crc16:
        raise ValueError("crc16 mismatch")

    if version != VERSION:
        raise ValueError(f"unexpected version: {version}")

    pkt = Packet(msg_type, zone_id, stream_id, seq, timestamp, flags, payload)
    if msg_type == MSG_AUDIO:
        pkt.payload_parsed = AudioPayload.parse(payload)
    elif msg_type == MSG_TELEMETRY:
        pkt.payload_parsed = TelemetryPayload.parse(payload)
    elif msg_type == MSG_CONTROL:
        pkt.payload_parsed = ControlPayload.parse(payload)
    else:
        pkt.payload_parsed = None

    return {"header": pkt, "payload": pkt.payload_parsed}


def _demo() -> None:
    aud = AudioPayload(
        encoding=ENC_PCM16_LE,
        channels=1,
        sample_rate_hz=48000,
        frame_samples=480,
        audio=bytes([0] * 960),
    )
    pkt = Packet(MSG_AUDIO, zone_id=2, stream_id=7, seq=1, timestamp=0, flags=FLAG_START, payload=aud.pack())
    raw = pkt.pack(include_crc=True)
    obj = parse_packet(raw)
    print(f"Demo packet bytes: {len(raw)}")
    print(f"Decoded: zone={obj['header'].zone_id}, seq={obj['header'].seq}, type={obj['header'].msg_type}")
    parsed_payload = obj["payload"]
    print(f"Payload samples={parsed_payload.frame_samples}, audio_bytes={len(parsed_payload.audio)}")


if __name__ == "__main__":
    _demo()
